Core/include/Acts/TrackFinder/CombinatorialKalmanFilter.hpp

```diff
@@ -0,0 +1,1255 @@
+          typename output_converter_t = VoidKalmanComponents>
+class CombinatorialKalmanFilter {
+ public:
+  /// Shorthand definition
+  using MeasurementSurfaces = std::multimap<const Layer*, const Surface*>;
+
+  /// Default constructor is deleted
+  CombinatorialKalmanFilter() = delete;
+
+  /// Constructor from arguments
+  CombinatorialKalmanFilter(
+      propagator_t pPropagator,
+      std::unique_ptr<const Logger> logger =
+          getDefaultLogger("CombinatorialKalmanFilter", Logging::INFO),
+      input_converter_t pInputCnv = input_converter_t(),
+      output_converter_t pOutputCnv = output_converter_t())
+      : m_propagator(std::move(pPropagator)),
+        m_inputConverter(std::move(pInputCnv)),
+        m_outputConverter(std::move(pOutputCnv)),
+        m_logger(logger.release()) {}
+
+ private:
+  /// The propgator for the transport and material update
+  propagator_t m_propagator;
+
+  /// The input converter to Fittable measurements
+  input_converter_t m_inputConverter;
+
+  /// The output converter into a given format
+  output_converter_t m_outputConverter;
+
+  /// Logger getter to support macros
+  const Logger& logger() const { return *m_logger; }
+
+  /// Owned logging instance
+  std::shared_ptr<const Logger> m_logger;
+
+  /// The navigator type
+  using KalmanNavigator = typename decltype(m_propagator)::Navigator;
+
+  /// The navigator has DirectNavigator type or not
+  static constexpr bool isDirectNavigator =
+      std::is_same<KalmanNavigator, DirectNavigator>::value;
+
+  /// @brief Propagator Actor plugin for the CombinatorialKalmanFilter
+  ///
+  /// @tparam source_link_t is an type fulfilling the @c SourceLinkConcept
+  /// @tparam parameters_t The type of parameters used for "local" paremeters.
+  ///
+  /// The CombinatorialKalmanFilterActor does not rely on the measurements to be
+  /// sorted along the track.
+  template <typename source_link_t, typename parameters_t>
+  class Actor {
+   public:
+    using TrackStateType = TrackState<source_link_t, parameters_t>;
+
+    /// Explicit constructor with updater and calibrator
+    Actor(updater_t pUpdater = updater_t(), smoother_t pSmoother = smoother_t(),
+          source_link_selector_t pSourceLinkSelector = source_link_selector_t(),
+          branch_stopper_t pBranchStopper = branch_stopper_t(),
+          calibrator_t pCalibrator = calibrator_t())
+        : m_updater(std::move(pUpdater)),
+          m_smoother(std::move(pSmoother)),
+          m_calibrator(std::move(pCalibrator)) {}
+
+    /// Broadcast the result_type
+    using result_type = CombinatorialKalmanFilterResult<source_link_t>;
+
+    /// Broadcast the track tip state type
+    using TipState = typename result_type::TipState;
+
+    /// The target surface
+    const Surface* targetSurface = nullptr;
+
+    /// Allows retrieving measurements for a surface
+    std::map<const Surface*, std::vector<source_link_t>> inputMeasurements;
```

Comment:
Why a map of vectors over a multimap?

Reply:
The selection of source links will be handled by the  `CKFSourceLinkSelector`.  All the source links on a surface will be passed to the `CKFSourceLinkSelector`, i.e. all the source links will be put into a container anyway. Also, the source links on a surface have to be somewhat indexed (since source links are not comparable) so that we could check if a source link has already been stored to avoid storage duplication. Does this make sense to you?

Reply:
Yeah I think I see the point.